server/game/cards/01_SOR/units/GuardianOfTheWhills.ts

```diff
@@ -0,0 +1,48 @@
+import AbilityHelper from '../../../AbilityHelper';
+import * as AbilityLimit from '../../../core/ability/AbilityLimit';
+import { Card } from '../../../core/card/Card';
+import { NonLeaderUnitCard } from '../../../core/card/NonLeaderUnitCard';
+import { StateWatcherRegistrar } from '../../../core/stateWatcher/StateWatcherRegistrar';
+import { CardsPlayedThisPhaseWatcher } from '../../../stateWatchers/CardsPlayedThisPhaseWatcher';
+
+export default class GuardianOfTheWhills extends NonLeaderUnitCard {
+    private cardsPlayedThisPhaseWatcher: CardsPlayedThisPhaseWatcher;
+
+    protected override getImplementationId() {
+        return {
+            id: '4166047484',
+            internalName: 'guardian-of-the-whills'
+        };
+    }
+
+    protected override setupStateWatchers(registrar: StateWatcherRegistrar): void {
+        this.cardsPlayedThisPhaseWatcher = AbilityHelper.stateWatchers.cardsPlayedThisPhase(registrar, this);
+    }
+
+    private isFirstUpgradePlayedOnThisCopy(card: Card, adjusterSource: Card): boolean {
+        if (!adjusterSource.isUnit()) {
+            return false;
+        }
+        const playedUpgradesOnThisCardThisPhase = this.cardsPlayedThisPhaseWatcher.getCardsPlayed((playedCardEntry) =>
+            playedCardEntry.card.isUpgrade() &&
+            playedCardEntry.card.parentCard === adjusterSource &&
```

Comment:
We can't directly check `card.parentCard` because the upgrade may have been attached to a new card since it was originally played. We need to add a new property to the watcher which captures the parent card that the card is attached to at the time it is first played. Once implemented, let's also add a test case for that using Survivor's Gauntlet

Reply:
Fixed and updated, now tracking parent card and the parent card's playedId at the time of the event as well